lib/git-metadata-preprocessor/extension.rb

```diff
@@ -0,0 +1,43 @@
+require 'asciidoctor/extensions' unless RUBY_ENGINE == 'opal'
+
+require 'rugged'
+
+class GitMetadataPreprocessor < Asciidoctor::Extensions::Preprocessor
+  def process document, reader
+
+    begin
+      repo = Rugged::Repository.discover(".")
+    rescue
+      $stderr.puts('Failed to find repository, git-metadata extension terminating')
+      exit
+    end
+
+    head = repo.head
+
+    document.attributes['git-metadata-sha'] = head.target_id
+    document.attributes['git-metadata-sha-short'] = head.target_id[0,8]
+    document.attributes['git-metadata-author-name'] = head.target.author[:name]
+    document.attributes['git-metadata-author-email'] = head.target.author[:email]
+    document.attributes['git-metadata-date-mdy'] = head.target.time.strftime("%m-%d-%y")
+    document.attributes['git-metadata-date-dmy'] = head.target.time.strftime("%d-%m-%y")
+    document.attributes['git-metadata-commit-message'] = head.target.message
+
+    begin
+      document.attributes['git-metadata-branch'] = repo.branches[head.name].name
+    rescue
+      document.attributes['git-metadata-branch'] = repo.branches[head.name]
```

Comment:
This line is a bit odd, could you please explain why we need to do that ?

Reply:
This is because I don't know enough about git. 

One of the reasons I was wanting this extension was to automatically build documentation and have the branch name flow into my document within a CI/CD environment. My CI tool (Gitlab) checks out the repository for pipelines in a detached head form and this was crashing the extension.

This "solved" that. I am not sure which checks I should make or how I should handle them. What are the possible repository states?
* empty (no commits yet, maybe we don't care about this, but it is still a valid case where there is a file but no commits. At present rugged provides a reference error in line 15). Should we exit the extension in this case?
* detached head (shouldn't cause the extension to crash). Should the branch say, for instance 'detached HEAD'?
* others?

The extension gives a None or empty string I think for a detached head. But how should we treat this. Should this unset the attribute instead? Same question for tags I guess.